### DISCUSSION TITLE
About a comprehensive resource on full waveform inversion code

### DISCUSSION THREAD

**Discussion by tjk9501** [slimgroup/JUDI.jl#231] [General]
Hello,
I have been using the Devito package to perform 2D as well as 3D waveform inversion in many different circumstances. And I have also tried many other different packages, which are written in C++, MATLAB, and many other different languages. I wonder if there are any open resources that can guide me through writing a complete full waveform inversion code from scratch (either a time domain or frequency domain guide), for studying and basic scientific research. Because in real applications, it will be necessary to integrate the complete code in various platforms, which might use a completely different language. Many previous published codes, which are written in various languages, are extremely hard to maintain and understand for beginners, not to mention rewriting and transferring to different platforms.

If you have any relevant resources, please notify me in the post, I will be happy to know and study the necessary resources.

**Comment by flexie** [comment]
Frankly, I am unsure why you would want to do that. The whole purpose of Devito is to avoid writing your own difficult to maintain code. Devito produces on-the-fly highly optimized parallel C codes that you may have a difficult time beating with hand-written code. Devito is being run in production at full industry scale, so why would you want to replicate that in what seems a multi-year development project. Instead, I would have a careful look how the Devito just-in-time compiler works.

Kind regards,

Felix

Felix J. Herrmann
Professor
EAS, CSE, ECE
GRA Eminent Scholar Chair in Energy
Seismic Laboratory for Imaging and Modeling

On Feb 14, 2024, at 11:20 AM, tjk9501 ***@***.***> wrote:



Hello,
I have been using the Devito package to perform 2D as well as 3D waveform inversion in many different circumstances. And I have also tried many other different packages, which are written in C++, MATLAB, and many other different languages. I wonder if there are any open resources that can guide me through writing a complete full waveform inversion code from scratch (either a time domain or frequency domain guide), for studying and basic scientific research. Because in real applications, it will be necessary to integrate the complete code in various platforms, which might use a completely different language. Many previous published codes, which are written in various languages, are extremely hard to maintain and understand for beginners, not to mention rewriting and transferring to different platforms.

If you have any relevant resources, please notify me in the post, I will be happy to know and study the necessary resources.

**Reply by mloubout** [reply]
Hi

Thanks for your interest. One of the main goals of JUDI is to be able to work at a level that makes it easy to teach and work with while still getting the performance needed to scale when moving to real-size problems.

There are quite a few notebooks in the documentation that go through modeling, IO, imaging, and inversion, and I think JUDI has all the features needed for teaching and learning about FWI. If some features are missing in the notebook, we gladly accept contributions and discussion.

You can also find a didactic FWI notebook that shows all the steps at

https://github.com/slimgroup/ConstrainedFWIExamples/blob/master/notebooks/01_constr_fwi_judi.ipynb

cheers

**Reply by tjk9501** [reply]
Thank you for your prompt response and generosity! I will very carefully read the codes you provide. By the way, I have to explain clearly about the first question that I post. In reality many software engineers don't use python or julia, so if it is possible we have to make the relevent packages even compatible with C or C++, because we have to be in tune with what everyone else is doing.

**Reply by mloubout** [reply]
While it is true that a lot of software engineers work in C/C++/Fortran, the vast majority of the practitioners, i.e., the geophysicists that run FWI, do not. This is one of the main reasons Devito is being widely adopted, as domain scientists cannot afford to spend years hand-coding new physics and algorithms in low-level language to try new ideas. Software, irrespective of the application, is to be designed for the user, not the developer. And yes, it is more work for the developer (this is what I do full-time), but it is what makes it viable.

Knowing how numerical methods work is good, I do agree. Still, for FWI software to be used, it has to be developed in a language and at an abstraction level that makes it accessible. This is also true for most graduate students who would lose a good half of their Ph.D. if they had to implement FWI from scratch every time. Our students can publish early results within their first year, if not 6 months, thanks to being able to focus on FWI and research and be set up within a week.
